### PR TITLE
[Core]Save task spec in separate table 

### PR DESCRIPTION
**Pull request by WangTaoTheTonic**
## Why are these changes needed?
This is a rebase version of https://github.com/ray-project/ray/pull/11592. As task spec info is only needed when gcs create or start an actor, so we can remove it from actor table and save the serialization time and memory/network cost when gcs clients get actor infos from gcs.

As internal repository varies very much from the community. This pr just add some manual check with simple cherry pick. Welcome to comment first and at the meantime I'll see if there's any test case failed or some points were missed.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
is it still WIP?

**Comment by ericl** [comment]
Hmm don't we need it for restarting failed actors?

**Comment by WangTaoTheTonic** [comment]
> is it still WIP?

Some dashboard test cases failed, I'll try to fix them today.

**Comment by WangTaoTheTonic** [comment]
> Hmm don't we need it for restarting failed actors?

We store task spec in separate table and get them when needed (like restarting failed actors).

**Comment by WangTaoTheTonic** [comment]
The failed windows tests and rllib tests are unrelated.

File: src/ray/protobuf/gcs.proto
Comment (rkooo567) [inline src/ray/protobuf/gcs.proto:34]: Can you just keep the prefix ordering and add `ACTOR_TASK_SPEC =21`?

**Review by rkooo567 — COMMENTED** [review]
I think it is a good optimization, but the new interface is a bit ugly (it is not clear from code when task spec is accessible or not). To make sure I understand, is this correct?

1. If the actor is dead, it wouldn't contain the task spec
2. Otherwise, it contains the task spec (reconstructing + creating a new actor)

**Comment by WangTaoTheTonic** [comment]
@rkooo567 Yeah. What you understand is correct. The task spec table acts like a foreign table of actor table, which stores some static informations.

File: src/ray/protobuf/gcs.proto
Comment (WangTaoTheTonic) [reply src/ray/protobuf/gcs.proto:34]: No problem. Putting it after `ACTOR` makes more readability as they are related. I'd like adding it just in the end. I don't have a strong preference here.

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:1139]:
```suggestion
  const auto &actor_task_specs = gcs_init_data.ActorTaskSpecs();
```

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:169]: Can you change this to `RAY_CHECK(state != DEAD)` and write a comment that task spec is not available when the actor is dead?

File: src/ray/gcs/gcs_server/gcs_actor_manager.h
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.h:166]: is it immutable?

File: src/ray/gcs/gcs_server/gcs_actor_manager.h
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.h:54]: Can you add RAY_CHECK(state != DEAD) in the body?

File: src/ray/gcs/gcs_server/gcs_actor_manager.h
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.h:49]: Can you update the docstring that this constructor is only for alive actors?

File: src/ray/gcs/gcs_server/gcs_actor_manager.h
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.h:93]: Can you clean up the task spec when the actor state becomes DEAD?

**Review by rkooo567 — COMMENTED** [review]
Mostly LGTM. For the GCS actor, I feel like the ideal interface is to hide task spec information (e.g., when it is obtained/removed) under `GcsActor` class, but it could be overkill. If you'd like to keep the current style, I just would like to make sure the invariants are guaranteed in the codebase (DEAD actor doesn't contain the task spec), and most of comments are about that!

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:1139]: ok

File: src/ray/gcs/gcs_server/gcs_actor_manager.h
Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_server/gcs_actor_manager.h:49]: ok

**Comment by WangTaoTheTonic** [comment]
@rkooo567 I've tried two ways to optimize code structure.
1. Add state check in `GetCreationTaskSpecification`, and reset `task_spec_` when its state being set as DEAD.
    The problem is that in many codes we still use `task_spec_` of `GcsActor` even we already set its state as DEAD, like https://github.com/ray-project/ray/blob/master/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L727, but we do clear `task_spec_` while destroying actor.
   The history makes code complicated, maybe we need another round to make them clean and more readable.
2. Tried to handle task spec table's addition and deletion in `GcsActor` class.
    Like you said, it's kinda overkill, because we need to pass `TaskSpecTable()` into `GcsActor` class while initialization. As `GcsActor` is only a wrapper of `ActorTableData`, this will make it complicated and probably not safe.

Other comments has been addressed.
About the two questions, if you have more good ideas I like to hear and discuss.

**Comment by rkooo567** [comment]
> Add state check in GetCreationTaskSpecification, and reset task_spec_ when its state being set as DEAD.

So this means this is not the right assumption https://github.com/ray-project/ray/pull/22650#pullrequestreview-898902342

Can you describe me exactly when this task spec is not available from GCSActor?

**Comment by rkooo567** [comment]
Actually, https://github.com/ray-project/ray/blob/678d23fe42196583a4996374d642feee78778c07/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L727

It looks like the task_id is only needed to cancel scheduling here.

Why don't we move this line to that else statement? https://github.com/ray-project/ray/blob/678d23fe42196583a4996374d642feee78778c07/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L769

**Comment by WangTaoTheTonic** [comment]
> Actually,
> 
> https://github.com/ray-project/ray/blob/678d23fe42196583a4996374d642feee78778c07/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L727
> 
> It looks like the task_id is only needed to cancel scheduling here.
> 
> Why don't we move this line to that else statement?
> 
> https://github.com/ray-project/ray/blob/678d23fe42196583a4996374d642feee78778c07/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L769

The assumption is right. I guess current implementation is not the best.
Moving the line to the else statement doesn't help, because the task speck is cleared in line https://github.com/ray-project/ray/blob/678d23fe42196583a4996374d642feee78778c07/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L728

**Comment by WangTaoTheTonic** [comment]
@rkooo567 
The check logic has been added and a tiny change has been applied to solve the test failures.
The lifecycle of actor and its `task_spec_` in memory and in redis table now consistent. As you said the fate sharing should be of more elegancy, I'd like to try to optimize them in future.

File: src/ray/gcs/gcs_server/gcs_actor_manager.h
Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_server/gcs_actor_manager.h]: it's already been cleared in https://github.com/ray-project/ray/blob/master/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L728
but i agree with you it should be more neat. I'd like to optimize them in future PRs.

File: src/ray/gcs/gcs_server/gcs_actor_manager.h
Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_server/gcs_actor_manager.h]: done

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc]: done

**Comment by rkooo567** [comment]
This sounds great! Thanks for trying to get this working! I will take a look at it very soon :)!!

**Comment by WangTaoTheTonic** [comment]
The failed tests seems flaky and unrelated

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:774]: I am a little concerned this will cause a random race condition. I think it is a pretty big change because at the end of the `DestroyActor`, some of states might not be properly updated.

If task_id is the only field that's required, why don't we actually just embed it into the actor table data? (only task id)

**Review by rkooo567 — COMMENTED** [review]
I have one last comment (I am a bit concerned about race condition since this code path can easily cause weird broken-up states)! Thanks for handling the comments!

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:774]: Putting task_id in actor table data is acceptable, but for long term, I would prefer to keep it clean.
Let me embed it for now, and remove it in future once we ensure the states being updated correctly.

Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc]: About `task_id` I found it's generated from actor id, like https://github.com/ray-project/ray/blob/master/src/ray/core_worker/core_worker.cc#L1737
So here is a simple(and stable) way for us to obtain task id, instead of getting it from any table.

**Review by rkooo567 — APPROVED** [review]
Thanks for addressing comments!

**Comment by rkooo567** [comment]
Has a cpp test failure. Can you check if it is related?

**Comment by WangTaoTheTonic** [comment]
> Has a cpp test failure. Can you check if it is related?

`object_lifecycle_manager_test` is about object creating, it's not related.

**Comment by WangTaoTheTonic** [comment]
@rkooo567 Is it ok to go?

File: src/ray/protobuf/gcs.proto
Comment (raulchen) [inline src/ray/protobuf/gcs.proto:164]: It seems that we don't have to put all these fields in the actor table. 
They are used for "GetNamedActorHandle". 
We can make GCS returns the actor table table + the task spec in the named actor query.

Comment (WangTaoTheTonic) [reply src/ray/protobuf/gcs.proto:153]: yeah that's brilliant!

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (raulchen) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:796]: There is an edge case where the actor is marked dead, but this delete doesn't finish. This will result in data leak. 
It's not a big deal, but we can check this case when loading data after GCS restarts.

**Review by raulchen — APPROVED** [review]

**Comment by rkooo567** [comment]
Sorry. I’ve been off (until this week). Please feel free to merge the pr! @raulchen

**Comment by rkooo567** [comment]
I am back. @WangTaoTheTonic I saw you added some new commits after it is approved. Is it ready to be merged? Also, can you add unit tests if those new code addition was due to some edge cases?

**Comment by WangTaoTheTonic** [comment]
> I am back. @WangTaoTheTonic I saw you added some new commits after it is approved. Is it ready to be merged? Also, can you add unit tests if those new code addition was due to some edge cases?

Yeah, it's ready to be merged.
I added some codes just to prevent the situation that the gcs crashes (no matter what reason) before the task spec is removing but the actor table item has been removed, per @raulchen comments. It's a small sanity check and won't affect the whole pr too much.
The unit test is hard to simulate that situation as we must kill gcs right in the duration between actor table item deleting and task spec item deleting.

**Comment by WangTaoTheTonic** [comment]
And the failed test cases are not related except the `LinkCheck`, @rkooo567 could you help to confirm?

**Comment by WangTaoTheTonic** [comment]
image
The failed reason of the two cases are same. It looks like the dashboard frontend issue and not related.

**Comment by WangTaoTheTonic** [comment]
@edoakes Hi Edward, this patch changes some proto structures in actor data table, so it impacts dashboard too. The change is only about messages but not api change. 
Could you help to take a review? As we need approvals of owners from modules involved.

File: dashboard/modules/actor/actor_head.py
Comment (edoakes) [inline dashboard/modules/actor/actor_head.py:65]: please delete `actor_classname_from_task_spec` if it's no longer used anywhere

File: dashboard/modules/actor/actor_head.py
Comment (edoakes) [inline dashboard/modules/actor/actor_head.py:70]: this looks like we're changing the returned schema -- we used to return `light_message["taskSpec"]["functionDescriptor"]` but now it's just `light_message["functionDescriptor"]`. We likely need to change the dashboard to account for this?

File: dashboard/modules/actor/actor_head.py
Comment (WangTaoTheTonic) [reply dashboard/modules/actor/actor_head.py:65]: We just don't use it here. It's still used by other codes.

File: dashboard/modules/actor/actor_head.py
Comment (WangTaoTheTonic) [reply dashboard/modules/actor/actor_head.py:70]: The dashboard uses `function descriptor` just for showing, we extract it from the returned schema, no matter in what position it is.

**Comment by rkooo567** [comment]
Btw, I need the approval from the protobuf code owner cc @edoakes @wuisawesome

**Review by edoakes — APPROVED** [review]

**Review by wuisawesome — APPROVED** [review]
the code owner parts lgtm, didn't review the rest

**Comment by rkooo567** [comment]
Failed test must be unrelated

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (simon-mo) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:378]: please remove this log line or make this more informational

Comment (WangTaoTheTonic) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:378]: @simon-mo Sorry for missing that, will remove it in https://github.com/ray-project/ray/pull/24204

**Pull request merged**
